File: src/widgets/status.zig
const std = @import("std");
const tui = @import("tui");
const Color = tui.Color;
const tmux_renderer = @import("../tmux_renderer.zig");
const themes = @import("../themes/registry.zig");
const Theme = @import("../core/theme.zig").Theme;
const cli = @import("../cli/args.zig");
const datetime = @import("datetime.zig");

const git_status = @import("git_status.zig");
const wb_git_status = @import("wb_git_status.zig");
const cpu_memory = @import("cpu_memory.zig");
const hostname = @import("hostname.zig");
const battery_mod = @import("battery.zig");
const kubernetes = @import("kubernetes.zig");
const cwd = @import("cwd.zig");
const terraform = @import("terraform.zig");
const docker_mod = @import("docker.zig");
const yadm = @import("yadm.zig");
const gpg_ssh_agent = @import("gpg_ssh_agent.zig");
const ai_assistant = @import("ai_assistant.zig");

const WidgetColor = enum {
    emphasis,
    success,
    accent,
    info,
    danger,
    accent_bright,
    primary,
    primary_bright,
    warning,
};

const WidgetEntry = struct {
    color: WidgetColor,
    no_sep: bool = false,
};

fn lookupEntry(name: []const u8) ?WidgetEntry {
    if (std.mem.eql(u8, name, "cwd")) return .{ .color = .emphasis };
    if (std.mem.eql(u8, name, "git")) return .{ .color = .success };
    if (std.mem.eql(u8, name, "wb-git")) return .{ .color = .accent, .no_sep = true };
    if (std.mem.eql(u8, name, "docker")) return .{ .color = .info, .no_sep = true };
    if (std.mem.eql(u8, name, "battery")) return .{ .color = .danger };
    if (std.mem.eql(u8, name, "hostname")) return .{ .color = .info };
    if (std.mem.eql(u8, name, "cpu")) return .{ .color = .accent_bright };
    if (std.mem.eql(u8, name, "kubernetes")) return .{ .color = .info };
    if (std.mem.eql(u8, name, "terraform")) return .{ .color = .primary };
    if (std.mem.eql(u8, name, "yadm")) return .{ .color = .accent };
    if (std.mem.eql(u8, name, "gpg-ssh")) return .{ .color = .primary_bright };
    if (std.mem.eql(u8, name, "ai-assistant")) return .{ .color = .success };
    if (std.mem.eql(u8, name, "datetime")) return .{ .color = .warning };
    return null;
}

fn colorFromTheme(theme: Theme, wc: WidgetColor) Color {
    return switch (wc) {
        .emphasis => theme.emphasis,
        .success => theme.success,
        .accent => theme.accent,
        .info => theme.info,
        .danger => theme.danger,
        .accent_bright => theme.accent_bright,
        .primary => theme.primary,
        .primary_bright => theme.primary_bright,
        .warning => theme.warning,
    };
}

fn renderWidget(
    arena: std.mem.Allocator,
    io: std.Io,
    environ_map: *std.process.Environ.Map,
    theme_name: []const u8,
    transparent: bool,
    pane_path: []const u8,
    widget_name: []const u8,
    battery_name: ?[]const u8,
    low_threshold: u8,
    time_format: cli.TimeFormat,
    cache_ttl: u64,
) ?[]const u8 {
    var buf: [4096]u8 = undefined;
    var writer: std.Io.Writer = .fixed(&buf);

    if (std.mem.eql(u8, widget_name, "cwd")) {
        cwd.run(arena, io, theme_name, transparent, environ_map, pane_path, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "git")) {
        git_status.run(arena, io, environ_map, theme_name, transparent, pane_path, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "wb-git")) {
        wb_git_status.run(arena, io, environ_map, theme_name, transparent, pane_path, cache_ttl, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "cpu")) {
        cpu_memory.run(arena, io, environ_map, theme_name, transparent, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "hostname")) {
        hostname.run(arena, io, theme_name, transparent, environ_map, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "datetime")) {
        datetime.run(arena, theme_name, time_format, transparent, io, environ_map, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "battery")) {
        battery_mod.run(arena, io, environ_map, theme_name, transparent, battery_name, low_threshold, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "kubernetes")) {
        kubernetes.run(arena, io, theme_name, transparent, environ_map, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "terraform")) {
        terraform.run(arena, io, theme_name, transparent, environ_map, pane_path, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "docker")) {
        docker_mod.run(arena, io, theme_name, transparent, environ_map, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "yadm")) {
        yadm.run(arena, io, theme_name, transparent, environ_map, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "gpg-ssh")) {
        gpg_ssh_agent.run(arena, io, environ_map, theme_name, transparent, &writer) catch return null;
    } else if (std.mem.eql(u8, widget_name, "ai-assistant")) {
        ai_assistant.run(arena, io, environ_map, theme_name, transparent, &writer) catch return null;
    } else return null;

    const output = std.Io.Writer.buffered(&writer);
    if (output.len == 0) return null;
    return arena.dupe(u8, output) catch null;
}

pub fn run(
    arena: std.mem.Allocator,
    io: std.Io,
    environ_map: *std.process.Environ.Map,
    theme_name: []const u8,
    transparent: bool,
    pane_path: ?[]const u8,
    show_names: []const []const u8,
    battery_name: ?[]const u8,
    low_threshold: u8,
    time_format_str: []const u8,
    cache_ttl: u64,
    writer: *std.Io.Writer,
) !void {
    const time_format = cli.TimeFormat.fromString(time_format_str) orelse .H24;
    const theme = (themes.byName(arena, io, environ_map, theme_name) orelse themes.hard).withTransparentBackground(transparent);
    const path = pane_path orelse ".";

    var outputs: std.ArrayList(struct { text: []const u8, color: WidgetColor, no_sep: bool }) = .empty;
    defer outputs.deinit(arena);

    for (show_names) |name| {
        const entry = lookupEntry(name) orelse continue;
        if (renderWidget(arena, io, environ_map, theme_name, transparent, path, name, battery_name, low_threshold, time_format, cache_ttl)) |output| {
            try outputs.append(arena, .{ .text = output, .color = entry.color, .no_sep = entry.no_sep });
        }
    }

    if (outputs.items.len == 0) return;

    var prev_no_sep: bool = false;
    for (outputs.items, 0..) |item, i| {
        if (i > 0 and !prev_no_sep and !item.no_sep) {
            try writer.writeAll(" ");
        }
        var fg_buf: [32]u8 = undefined;
        var bg_buf: [32]u8 = undefined;
        try writer.print("#[fg={s},bg={s}]", .{ tmux_renderer.colorHexString(colorFromTheme(theme, item.color), &fg_buf), tmux_renderer.colorHexString(theme.surface_alt, &bg_buf) });
        try writer.writeAll(item.text);
        prev_no_sep = item.no_sep;
    }
}

test "lookupEntry returns correct colors for all widget names" {
    const cases = [_]struct { name: []const u8, color: WidgetColor, no_sep: bool }{
        .{ .name = "cwd", .color = .emphasis, .no_sep = false },
        .{ .name = "git", .color = .success, .no_sep = false },
        .{ .name = "wb-git", .color = .accent, .no_sep = true },
        .{ .name = "docker", .color = .info, .no_sep = true },
        .{ .name = "battery", .color = .danger, .no_sep = false },
        .{ .name = "hostname", .color = .info, .no_sep = false },
        .{ .name = "cpu", .color = .accent_bright, .no_sep = false },
        .{ .name = "kubernetes", .color = .info, .no_sep = false },
        .{ .name = "terraform", .color = .primary, .no_sep = false },
        .{ .name = "yadm", .color = .accent, .no_sep = false },
        .{ .name = "gpg-ssh", .color = .primary_bright, .no_sep = false },
        .{ .name = "ai-assistant", .color = .success, .no_sep = false },
        .{ .name = "datetime", .color = .warning, .no_sep = false },
    };

    for (cases) |c| {
        const entry = lookupEntry(c.name) orelse return error.TestFailed;
        try std.testing.expectEqual(c.color, entry.color);
        try std.testing.expectEqual(c.no_sep, entry.no_sep);
    }

    try std.testing.expectEqual(@as(@TypeOf(lookupEntry("")), null), lookupEntry("nonexistent"));
}

test "colorFromTheme maps WidgetColor to correct Theme fields" {
    const theme = themes.hard;
    try std.testing.expect(std.meta.eql(theme.emphasis, colorFromTheme(theme, .emphasis)));
    try std.testing.expect(std.meta.eql(theme.success, colorFromTheme(theme, .success)));
    try std.testing.expect(std.meta.eql(theme.accent, colorFromTheme(theme, .accent)));
    try std.testing.expect(std.meta.eql(theme.info, colorFromTheme(theme, .info)));
    try std.testing.expect(std.meta.eql(theme.danger, colorFromTheme(theme, .danger)));
    try std.testing.expect(std.meta.eql(theme.accent_bright, colorFromTheme(theme, .accent_bright)));
    try std.testing.expect(std.meta.eql(theme.primary, colorFromTheme(theme, .primary)));
    try std.testing.expect(std.meta.eql(theme.primary_bright, colorFromTheme(theme, .primary_bright)));
    try std.testing.expect(std.meta.eql(theme.warning, colorFromTheme(theme, .warning)));
}

test "status run produces empty output for empty widget list" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var buf: [256]u8 = undefined;
    var writer: std.Io.Writer = .fixed(&buf);

    try run(gpa, io, &env_map, "hard", false, null, &.{}, null, 20, "24H", 300, &writer);
    try std.testing.expectEqual(@as(usize, 0), std.Io.Writer.buffered(&writer).len);
}

test "status run handles unknown widget names gracefully" {
    const gpa = std.testing.allocator;
    const io = std.Io.threaded_global.ioBasic();
    var env_map = std.process.Environ.Map.init(gpa);
    defer env_map.deinit();

    var buf: [256]u8 = undefined;
    var writer: std.Io.Writer = .fixed(&buf);

    try run(gpa, io, &env_map, "hard", false, null, &.{ "nonexistent", "also-fake" }, null, 20, "24H", 300, &writer);
    try std.testing.expectEqual(@as(usize, 0), std.Io.Writer.buffered(&writer).len);
}

File: src/widgets/status.rs
use crate::cli::args::TimeFormat;
use crate::core::Theme;
use crate::tmux_renderer::ThemeHex;
use crate::widgets;
use crate::widgets::status_entries::{color_from_theme, lookup_entry, WidgetColor, WidgetEntry};
use std::thread;

/// Shared config passed to widget rendering functions.
#[derive(Clone, Copy)]
pub struct WidgetConfig<'a> {
    pub theme: Theme,
    pub transparent: bool,
    pub pane_path: &'a str,
    pub battery_name: Option<&'a str>,
    pub low_threshold: u8,
    pub time_format: TimeFormat,
    pub cache_ttl: u64,
    pub separator: &'a str,
}

fn resolve_separator(style: &str) -> &'static str {
    match style {
        "pipe" => "│",
        "chevron" => "〉",
        "arrow" => "▸",
        "slash" => "/",
        "line" => "━",
        "block" => "▏",
        "none" => "",
        _ => " ",
    }
}

fn render_widget(cfg: WidgetConfig<'_>, theme_hex: &ThemeHex, widget_name: &str) -> Option<String> {
    let output = match widget_name {
        "cwd" => widgets::cwd::run_with_theme_hex(cfg.theme, theme_hex, cfg.pane_path),
        "git" => {
            // The porcelain is prefetched once per render (see `run`) so the
            // git widget never spawns its own `git status` — the whole render
            // shares one spawn for the repo.
            let status = widgets::git_status::prefetch_porcelain(cfg.pane_path);
            widgets::git_status::run_with_theme_hex_prefetched(
                cfg.theme,
                theme_hex,
                &status,
                cfg.pane_path,
            )
        }
        "wb-git" => widgets::wb_git_status::run_with_theme_hex(
            cfg.theme,
            theme_hex,
            cfg.pane_path,
            cfg.cache_ttl,
        ),
        "cpu" => widgets::cpu_memory::run_with_theme_hex(cfg.theme, theme_hex),
        "datetime" => widgets::datetime::run_with_theme_hex(cfg.theme, theme_hex, cfg.time_format),
        "battery" => widgets::battery::run_with_theme_hex(
            cfg.theme,
            theme_hex,
            cfg.battery_name,
            cfg.low_threshold,
        ),
        "kubernetes" => widgets::kubernetes::run_with_theme_hex(cfg.theme, theme_hex),
        "gpg-ssh" => widgets::gpg_ssh_agent::run_with_theme_hex(cfg.theme, theme_hex),
        _ => return None,
    };

    if output.is_empty() {
        None
    } else {
        Some(output)
    }
}

/// Assemble the full tmux status-right string from multiple widgets.
/// Each widget output is wrapped with its assigned color on the segment background.
pub fn run(cfg: WidgetConfig<'_>, show_names: &[&str]) -> String {
    let theme = cfg.theme;
    let theme_hex = ThemeHex::from_theme(theme);
    let segment_bg = if cfg.transparent {
        theme.surface_alt
    } else {
        theme.surface
    };

    struct WidgetOutput {
        text: String,
        color: WidgetColor,
        no_sep: bool,
    }

    let jobs: Vec<(&str, WidgetEntry)> = show_names
        .iter()
        .filter_map(|&name| lookup_entry(name).map(|entry| (name, entry)))
        .collect();
    // Position-indexed so parallel results land in their requested slot —
    // otherwise blocking widgets (git/forge) would always render last.
    let mut outputs: Vec<Option<WidgetOutput>> = (0..jobs.len()).map(|_| None).collect();

    // Most widgets are cache-read + format (~0.2ms); only the forge and
    // git widgets spawn subprocesses. Spawning a thread per widget costs
    // ~10-30µs each and only pays off when several subprocess-backed widgets
    // run concurrently. Batch the cheap widgets inline, parallelize only the
    // ones that actually block on external processes.
    let blocking = ["git", "wb-git"];
    let mut blocking_slots: Vec<usize> = Vec::new();

    for (i, (name, entry)) in jobs.iter().enumerate() {
        if blocking.contains(name) {
            blocking_slots.push(i);
            continue;
        }
        if let Some(text) = render_widget(cfg, &theme_hex, name) {
            outputs[i] = Some(WidgetOutput {
                text: text.trim_end().to_owned(),
                color: entry.color,
                no_sep: entry.no_sep,
            });
        }
    }

    if !blocking_slots.is_empty() {
        thread::scope(|scope| {
            let handles: Vec<_> = blocking_slots
                .iter()
                .map(|&slot| {
                    let (name, entry) = jobs[slot];
                    let theme_hex_ref = &theme_hex;
                    scope.spawn(move || {
                        let out = render_widget(cfg, theme_hex_ref, name).map(|text| {
                            WidgetOutput {
                                text: text.trim_end().to_owned(),
                                color: entry.color,
                                no_sep: entry.no_sep,
                            }
                        });
                        (slot, out)
                    })
                })
                .collect();

            for handle in handles {
                if let Ok((slot, Some(output))) = handle.join() {
                    outputs[slot] = Some(output);
                }
            }
        });
    }

    // Drop empty slots, preserving the requested relative order.
    let outputs: Vec<WidgetOutput> = outputs.into_iter().flatten().collect();
    if outputs.is_empty() {
        return String::new();
    }

    let sep = resolve_separator(cfg.separator);
    // "space" and "none" separators emit no glyph — the per-segment padding
    // provides the inter-segment gap. Glyph separators (pipe/chevron/arrow/
    // slash/line/block) sit between the padded segments, one space each side
    // from the padding.
    let sep_glyph = if matches!(cfg.separator, "space" | "none") {
        ""
    } else {
        sep
    };
    let mut result = String::with_capacity(1024);
    let mut prev_no_sep = false;

    for (i, item) in outputs.iter().enumerate() {
        let is_last = i + 1 == outputs.len();
        // no_sep widgets (forge) attach to the previous segment — no
        // glyph and no gap before them.
        let next_no_sep = outputs.get(i + 1).map(|n| n.no_sep).unwrap_or(false);

        if i > 0 && !prev_no_sep && !item.no_sep {
            result.push_str(sep_glyph);
        }

        // Each widget is its own segment on the segment background. One
        // trailing space pads the segment so widgets read as distinct blocks
        // instead of a run-on blob; the last segment omits it so the bar's
        // right edge stays clean (a padded space would paint a segment-bg
        // cell at the window edge).
        result.push_str(&format!(
            "#[fg={},bg={}]{}",
            theme_hex.color(color_from_theme(theme, item.color)),
            theme_hex.color(segment_bg),
            item.text,
        ));
        if !is_last && !next_no_sep {
            result.push(' ');
        }
        prev_no_sep = item.no_sep;
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::themes;

    #[test]
    fn status_run_produces_empty_output_for_empty_widget_list() {
        let cfg = WidgetConfig {
            theme: themes::HARD,
            transparent: false,
            pane_path: ".",
            battery_name: None,
            low_threshold: 20,
            time_format: TimeFormat::H24,
            cache_ttl: 300,
            separator: "space",
        };
        let output = run(cfg, &[]);
        assert!(output.is_empty());
    }

    #[test]
    fn status_run_handles_unknown_widget_names_gracefully() {
        let cfg = WidgetConfig {
            theme: themes::HARD,
            transparent: false,
            pane_path: ".",
            battery_name: None,
            low_threshold: 20,
            time_format: TimeFormat::H24,
            cache_ttl: 300,
            separator: "space",
        };
        let output = run(cfg, &["nonexistent", "also-fake"]);
        assert!(output.is_empty());
    }

    #[test]
    fn resolve_separator_returns_bare_glyphs() {
        // Separator glyphs are bare; spacing is applied in the assembly loop
        // so block/line styles can sit tight against widget text.
        assert_eq!("〉", resolve_separator("chevron"));
        assert_eq!("│", resolve_separator("pipe"));
        assert_eq!("▸", resolve_separator("arrow"));
        assert_eq!("/", resolve_separator("slash"));
        assert_eq!("━", resolve_separator("line"));
        assert_eq!("▏", resolve_separator("block"));
        assert_eq!("", resolve_separator("none"));
    }

    #[test]
    fn segments_are_padded_and_separated() {
        // cpu (no_sep=false) and datetime (no_sep=false): the default
        // space separator emits no glyph; the per-segment padding provides
        // the gap, and the last segment has no trailing space.
        let cfg = WidgetConfig {
            theme: themes::HARD,
            transparent: false,
            pane_path: ".",
            battery_name: None,
            low_threshold: 20,
            time_format: TimeFormat::H24,
            cache_ttl: 300,
            separator: "space",
        };
        let output = run(cfg, &["cpu", "datetime"]);
        // cpu and datetime both render. Assert the structural
        // properties: each segment is wrapped in fg/bg, exactly one space
        // separates them, and the output does not end with a space.
        assert!(output.contains("#[fg="));
        assert!(output.contains("bg="));
        assert!(!output.ends_with(' '));
        // The cpu segment's padding space sits before the datetime
        // segment's fg wrapper — the cpu text (the mem `%` value) is
        // followed by " " then a fresh "#[fg=".
        assert!(
            output.contains(" #[fg="),
            "segment padding space must precede the next segment wrapper: {output}"
        );
        // datetime (the last segment) has no trailing padding space — the
        // output must not end with a space.
        assert!(!output.ends_with(' '));
    }

    #[test]
    fn all_widgets_are_separate_segments() {
        // git and wb-git are both normal (no_sep=false) segments — the forge
        // widget renders with a space gap after the git segment, like every
        // other widget.
        let cfg = WidgetConfig {
            theme: themes::HARD,
            transparent: false,
            pane_path: ".",
            battery_name: None,
            low_threshold: 20,
            time_format: TimeFormat::H24,
            cache_ttl: 300,
            separator: "space",
        };
        let output = run(cfg, &["git", "wb-git"]);
        // In this repo both render; the forge header (muted icon) must sit
        // in its own segment after the git text.
        if output.contains("󱓎") && output.contains("") {
            let git_end = output.find("󱓎").unwrap();
            let forge_start = output.find("").unwrap();
            assert!(
                forge_start > git_end,
                "forge should come after git sync icon"
            );
            // The git segment's padding space sits before the forge segment's
            // outer wrapper (fg=accent,bg=surface), separating the two.
            let git_text_end = output.find("#[fg=#d3869b,bg=#282828]").unwrap();
            assert!(
                output[..git_text_end].ends_with(' '),
                "git segment must end with a padding space: '{output}'"
            );
        }
    }
}
